File: proto-csharp/ProtoMessage/ProtoMessage/ApeMessage/Patch/DiffPair.cs
using System;
using com.spaceape.utils;

namespace com.spaceape.protobuf.patch
{
  public class DiffPair
  {
    public String left{ get; private set; }

    public String right{ get; private set; }

    public DiffPair(String left, String right)
    {
      this.left = left;
      this.right = right;
    }

    public override Boolean Equals (Object obj){
        if (obj == null) return false;
        if (!(obj is DiffPair)) return false;
        DiffPair o = (DiffPair)obj;
        return left == o.left && right == o.right;
    }

    public override int GetHashCode(){
        return new HashCodeBuilder().Add(left).Add(right).GetHashCode();
    }
  }
}

File: proto-java/src/main/java/com/spaceape/protobuf/patch/DiffPair.java
package com.spaceape.protobuf.patch;

import org.apache.commons.lang3.builder.HashCodeBuilder;

public class DiffPair {
    public final String left;
    public final String right;
    public DiffPair(String left, String right){
        this.left = left;
        this.right = right;
    }

    @Override
    public boolean equals (Object obj){
        if (obj == null) return false;
        if (!(obj instanceof DiffPair)) return false;
        DiffPair o = (DiffPair)obj;
        return left == o.left && right == o.right;
    }

    @Override
    public int hashCode(){
        return new HashCodeBuilder().append(left).append(right).hashCode();
    }
}
